src/peft/tuners/oft/layer.py

```diff
@@ -0,0 +1,398 @@
+                    )
+
+                base_layer.weight.data = orig_weights
+                self.merged_adapters.append(active_adapter)
+
+    def unmerge(self) -> None:
+        if not self.merged:
+            warnings.warn("Already unmerged. Nothing to do.")
+            return
+        while len(self.merged_adapters) > 0:
+            active_adapter = self.merged_adapters.pop()
+            if active_adapter in self._available_adapters:
+                base_layer = self.base_layer
+                orig_weights = base_layer.weight.data
+                if isinstance(base_layer, nn.Linear):
+                    orig_weights = torch.transpose(orig_weights, 0, 1)
+                elif isinstance(base_layer, nn.Conv2d):
+                    orig_weights = orig_weights.view(
+                        [
+                            base_layer.kernel_size[0] * base_layer.kernel_size[1] * base_layer.in_channels,
+                            base_layer.out_channels,
+                        ]
+                    )
+                delta_weight = self.get_delta_weight(active_adapter)
+                if orig_weights.shape[0] != delta_weight.shape[1]:
+                    # when in channels is not divisible by r
+                    delta_weight = delta_weight[: orig_weights.shape[0], : orig_weights.shape[0]]
+                orig_weights = torch.linalg.solve(delta_weight, orig_weights)
+                if isinstance(base_layer, nn.Linear):
+                    orig_weights = torch.transpose(orig_weights, 0, 1)
+                elif isinstance(base_layer, nn.Conv2d):
+                    orig_weights = orig_weights.reshape(
+                        [
+                            base_layer.out_channels,
+                            base_layer.in_channels,
+                            base_layer.kernel_size[0],
+                            base_layer.kernel_size[1],
+                        ]
+                    )
+                self.base_layer.weight.data = orig_weights
+
+    def get_delta_weight(self, adapter_name: str) -> torch.Tensor:
+        scaling = self.scaling[adapter_name]
+        rank = self.r[adapter_name]
+        coft = self.coft[adapter_name]
+        eps = self.eps[adapter_name]
+        opt_r = self.oft_r[adapter_name]
+
+        if coft:
+            with torch.no_grad():
+                opt_r.copy_(self._project_batch(opt_r, eps=eps))
+
+        orth_rotate = self._cayley_batch(opt_r * scaling)
+        weight = self._block_diagonal(orth_rotate, rank)
+
+        return weight
+
+    def _cayley_batch(self, data: torch.Tensor) -> torch.Tensor:
+        b, r, c = data.shape
+        # Ensure the input matrix is skew-symmetric
+        skew = 0.5 * (data - data.transpose(1, 2))
+        I = torch.eye(r, device=data.device).unsqueeze(0).expand(b, r, c)
+
+        # Perform the Cayley parametrization
+        Q = torch.bmm(I - skew, torch.inverse(I + skew))
```

Comment:
https://github.com/Zeju1997/oft/blob/main/oft-t2i/oft_utils/attention_processor.py#L505

The official implementation uses inverse, which is why I use it.
However, I understand your opinion, and I will train models using `pinv`. If it works, we can add `pinv` option.

Reply:
Nice. Probably the authors had a good reason to use exact inverse and not pinv, but it could be worth a try.

Reply:
When I replaced `torch.inverse` with `torch.pinverse`, it raised the error `RuntimeError: expected scalar type Float but found Half`.  To avoid this error, I tried disabling autocast by setting `autocast enbale=False`. However, this resulted in longer training times compared to using `torch.inverse`. Therefore, I believe there is no need to support `pinv`.

```
if pinv and torch.is_autocast_enabled():
    device = 'cuda' if data.is_cuda else 'cpu'
    # to avoid `RuntimeError: expected scalar type Float but found Half`
    with torch.autocast(device_type=device, enabled=False):
        Q = torch.bmm(I - skew, torch.pinverse(I + skew))
else:
    Q = torch.bmm(I - skew, torch.inverse(I + skew))
```

Reply:
Thanks for testing, I wasn't aware that `pinv` doesn't work with half precision. As you wrote, it doesn't make sense then.

Something I wonder though is why get an issue with half precision in the first place. The `I` tensor that is created is in float32, which means that `I + skew` is also upcast to float32, even if `data` is half precision. How is it possible to get the runtime error then when using `pinv`?

Also, in my testing, `torch.inverse` also requires full precision:

```python
torch.inverse(torch.rand(10, 10).half())
# => RuntimeError: linalg.inv: Low precision dtypes not supported. Got Half
```

Any idea what could explain this?

Reply:
Hmm...
`torch.inverse` works well with `torch.autocast`. However, I don't have any ideas to explain it.

Reply:
And the error was raised in the backward process.

```
 File "/usr/local/lib/python3.10/dist-packages/mmengine/optim/optimizer/optimizer_wrapper.py", line 196, in update_params
    self.backward(loss)
  File "/usr/local/lib/python3.10/dist-packages/mmengine/optim/optimizer/amp_optimizer_wrapper.py", line 125, in backward
    self.loss_scaler.scale(loss).backward(**kwargs)
  File "/usr/local/lib/python3.10/dist-packages/torch/_tensor.py", line 491, in backward
    torch.autograd.backward(
  File "/usr/local/lib/python3.10/dist-packages/torch/autograd/__init__.py", line 204, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: expected scalar type Float but found Half
```